File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[][] prefixSum = new int[N][N];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                prefixSum[i][j] = Integer.parseInt(st.nextToken());
                if (i != 0) {
                    prefixSum[i][j] += prefixSum[i - 1][j];
                }
                if (j != 0) {
                    prefixSum[i][j] += prefixSum[i][j - 1];
                }
                if (i != 0 && j != 0) {
                    prefixSum[i][j] -= prefixSum[i - 1][j - 1];
                }
            }
        }

        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int x1 = Integer.parseInt(st.nextToken()) - 1;
            int y1 = Integer.parseInt(st.nextToken()) - 1;
            int x2 = Integer.parseInt(st.nextToken()) - 1;
            int y2 = Integer.parseInt(st.nextToken()) - 1;

            int result = prefixSum[x2][y2];
            if (x1 > 0) {
                result -= prefixSum[x1 - 1][y2];
            }
            if (y1 > 0) {
                result -= prefixSum[x2][y1 - 1];
            }
            if (x1 > 0 && y1 > 0) {
                result += prefixSum[x1 - 1][y1 - 1];
            }
            sb.append(result).append('\n');
        }

        System.out.print(sb);
        br.close();
    }
}

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.kt
import java.io.*
import java.util.StringTokenizer

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val insert1 = StringTokenizer(readLine())
    val N = insert1.nextToken().toInt()
    val M = insert1.nextToken().toInt()

    val sumOfIntervals = Array(N) { LongArray(N) }
    repeat(N) { i ->
        val insert2 = StringTokenizer(readLine())
        repeat(N) { j ->
            if (i == 0 && j == 0) {
                sumOfIntervals[i][j] = insert2.nextToken().toLong()
            } else if (i == 0) {
                sumOfIntervals[i][j] = sumOfIntervals[i][j - 1] + insert2.nextToken().toLong()
            } else if (j == 0) {
                sumOfIntervals[i][j] = sumOfIntervals[i - 1][j] + insert2.nextToken().toLong()
            } else {
                sumOfIntervals[i][j] =
                    sumOfIntervals[i - 1][j] + sumOfIntervals[i][j - 1] - sumOfIntervals[i - 1][j - 1] + insert2.nextToken().toLong()
            }
        }
    }

    val stringBuilder = StringBuilder()
    repeat(M) {
        val insert3 = StringTokenizer(readLine())
        val x1 = insert3.nextToken().toInt() - 1
        val y1 = insert3.nextToken().toInt() - 1
        val x2 = insert3.nextToken().toInt() - 1
        val y2 = insert3.nextToken().toInt() - 1

        val T = sumOfIntervals[x2][y2]
        val S1 = if (x1 - 1 >= 0) sumOfIntervals[x1 - 1][y2] else 0
        val S2 = if (y1 - 1 >= 0) sumOfIntervals[x2][y1 - 1] else 0
        val S3 = if (x1 - 1 >= 0 && y1 - 1 >= 0) sumOfIntervals[x1 - 1][y1 - 1] else 0
      stringBuilder.appendLine(T - S1 - S2 + S3)
    }

    print(stringBuilder.toString())
}